File: python/0003.py
import math
import numpy as np # for array

def primes(U):
    sieve = np.full(U, True, dtype=bool)
    sieve[0] = sieve[1] = False

    for i in range(U):
        if not sieve[i]: continue
        sieve[i*i::i] = False

        yield i

# math.isqrt is included from 3.8+
isqrt = lambda n: math.floor(math.sqrt(n))

L = 600_851_475_143
P = 1

# This is a bad solution
for p in primes(isqrt(L)):
    if L < p**2: break
    if L % p != 0: continue
    P = p
    while L % p == 0: L //= p

solution3 = max(L, P)

print(solution3)

File: haskell/0003.hs
main = putStrLn $ show largestprimedivisor

bignum = 600851475143


-- shamelessly taken from haskell.org frontpage
-- really slow for p > 1_000_000 (~5 min)
primes = filterPrime [2..]
    where filterPrime (p:xs) =
            -- each new prime adds a new x `mod` p /= 0 condition
            p : filterPrime [x | x <- xs, x `mod` p /= 0]


-- this is techincally gives the largest factor below the square root, so this solution doesn't work for squarefree semiprimes.
largestprimedivisor = maximum $ filter (`divides` bignum) (takeWhile (< isqrt bignum) primes)
    where
        divides :: Int -> Int -> Bool
        x `divides` y = y `mod` x == 0

        isqrt :: Int -> Int
        isqrt = floor . sqrt . fromIntegral